File: src/c/hmm.c
// Hidden Markov Model
// Author: WeiChung Huang
// Last Modified on 9/6/2017

#include "stdio.h"
#include "stdlib.h"
#include "time.h"
#include "math.h"
#include <ctype.h>
#include <string.h>

#include "mycipher.h"

void hmm_train(int M, int N, int T, int *O, int max_iter);
void init_model(int M, int N, int update_A);
void dump_model(int M, int N);
void alpha_pass(int N, int T, int *O);
void alpha_pass_no_scale(int N, int T, int *O);
void beta_pass(int N, int T, int *O);
void compute_gamma(int N, int T, int *O);
void re_estimate(int N, int M, int T, int *O);
double compute_prob(int T);
void init_brown_corpus(int T);
void init_simple_corpus(int T);
double get_all_states_prob_sum(int N, int T, int *O, int i, int t, double prev_prob);
double get_all_observations_prob_sum(int M, int N, int T, int t, int *O, int use_alpha_pass);
int verify_all_prob_is_valid(int M, int N, int T, int *O);

double **A = NULL;
double **B = NULL;
double *pi = NULL;

int *O = NULL;
double *C = NULL;

double **alpha = NULL;
double **beta = NULL;
double **gamma_t = NULL;
double ***di_gamma = NULL;

int g_random_seed = 0;

void dump_model(int M, int N) {
    // dump A-matrix
    printf("\n--------A-Matrix--------\n");
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%.4f ", A[i][j]);
        }
        printf("\n");
    }

    printf("\n--------B-Matrix--------\n");
    for (int i = 0; i < M; i++) {
        // if (i < M - 1) printf("    %c | ", i + 'a');
        // else printf("space | ");
        printf("    %c | ", i + 'a');
        for (int j = 0; j < N; j++) {
            printf("%.4f ", B[j][i]);
        }
        printf("\n");
    }

    printf("\n--------pi-Matrix--------\n");
    for (int i = 0; i < N; i++) {
        printf("%.4f ", pi[i]);
    }
    printf("\n");
}

void init_simple_corpus(int T) {
    if (T != 4) {fprintf(stderr, "ERROR: wrong T value."); exit(1);};

    A[0][0] = 0.7; A[0][1] = 0.3;
    A[1][0] = 0.4; A[1][1] = 0.6;

    B[0][0] = 0.1; B[0][1] = 0.4; B[0][2] = 0.5;
    B[1][0] = 0.7; B[1][1] = 0.2; B[1][2] = 0.1;

    pi[0] = 0.6; pi[1] = 0.4;

    O = (int *)malloc(sizeof(int) * T);
    O[0] = 0; O[1] = 1;
    O[2] = 0; O[3] = 2;
}

void init_brown_corpus(int T) {
    // init observations
    O = (int *)malloc(sizeof(int) * T);

    char * line = NULL;
    size_t len = 0;
    FILE *fp = fopen("../dataset/brown.txt", "r");
    if (fp == NULL) {fprintf(stderr, "ERROR: Couldn't open file (brown.txt).\n"); exit(1);};

    ssize_t read;
    int count = 0;
    while ((read = getline(&line, &len, fp)) != -1){
        for (int i = 0; i < len; i++) {
            char c = line[i];
            if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == ' ') {
                if (c != ' ') O[count] = tolower(c) - 'a';
                else O[count] = 26;
                count++;
            }
            if (count >= T) break;
        }
        if (count >= T) break;
    }
    if (count < T) exit(0);    
}

void init_ss_cipher_corpus(int T) {
    char *cipher = gen_cipher_with_shifting(T);
    
    O = (int *)malloc(sizeof(int) * T);
    for (int i = 0; i < cipher_length; i++) O[i] = cipher[i] - 'a';
    free(cipher);

    double **digraph = init_english_digraph();
    for (int i = 0; i < 26; i++) {
        for (int j = 0; j < 26; j++) A[i][j] = digraph[i][j];
    }
    free(digraph);
}

double get_all_states_prob_sum(int N, int T, int *O, int i, int t, double prev_prob) {
    if (t >= T) {
        return prev_prob;
    }

    double prob = 0.0;
    double curr_prob = 0.0;
    for (int j = 0; j < N; j++) {
        if (t == 0) curr_prob = pi[j] * B[j][O[t]];
        else curr_prob = prev_prob * (A[i][j] * B[j][O[t]]);

        prob += get_all_states_prob_sum(N, T, O, j, t + 1, curr_prob);

        curr_prob = prev_prob;
    }
    return prob;
}

double get_all_observations_prob_sum(int M, int N, int T, int t, int *O, int use_alpha_pass) {
    if (t >= T) {
        double sum = 0.0;
        if (use_alpha_pass) {
            alpha_pass_no_scale(N, T, O);
            for (int i = 0; i < N; i++) sum += alpha[i][T - 1];
            if (alpha) {
                for (int i = 0; i < N; i++) free(alpha[i]);
                free(alpha);
            }
        } else {
            sum = get_all_states_prob_sum(N, T, O, 0, 0, 1.0);
        }
        return sum;
    }

    double sum = 0.0;
    for (int i = 0; i < M; i++) {
        O[t] = i;
        sum += get_all_observations_prob_sum(M, N, T, t + 1, O, use_alpha_pass);
    }
    return sum;
}

int verify_all_prob_is_valid(int M, int N, int T, int *O) {
    // verify all probability are sum to one
    // method 1: using original method
    double sum1 = get_all_observations_prob_sum(M, N, T, 0, O, 0);
    // method 2: using alpha-pass
    double sum2 = get_all_observations_prob_sum(M, N, T, 0, O, 1);

    return (fabs(sum1 - 1.0) < 10e-7) && (fabs(sum2 - 1.0) < 10e-7);
}

void free_global_memory(int N) {
    if (A) {
        for (int i = 0; i < N; i++) free(A[i]);
        free(A); A = NULL;
    }
    if (B) {
        for (int i = 0; i < N; i++) free(B[i]);
        free(B); B = NULL;
    }
    if (pi) {free(pi); pi = NULL;}
    if (O) {free(O); O = NULL;}
    if (C) {free(C); C = NULL;}

    // free memory
    if (alpha) {
        for (int i = 0; i < N; i++) free(alpha[i]);
        free(alpha);
    }

    if (beta) {
        for (int i = 0; i < N; i++) free(beta[i]);
        free(beta);
    }

    if (gamma_t) {
        for (int i = 0; i < N; i++) free(gamma_t[i]);
        free(gamma_t);   
    }

    if (di_gamma) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                free(di_gamma[i][j]);
            }
            free(di_gamma[i]);
        }
        free(di_gamma);
    } 
}

int main(int argc, char *argv[]) {
    if (argc < 6) {
        fprintf(stderr, "[Usage]: M N T maxIteration numModels seed(optional)\n");
        exit(1);
    }
    const int M = atoi(argv[1]);
    const int N = atoi(argv[2]);
    const int T = atoi(argv[3]);
    const int max_iter = atoi(argv[4]);
    const int num_models = atoi(argv[5]);
    g_random_seed = time(NULL);
    if (argc > 6) g_random_seed = atoi(argv[6]);

    float max_score = 0.f;
    init_model(M, N, 1);
    init_ss_cipher_corpus(T);
    for (int m = 0; m < num_models; m++) {
        hmm_train(M, N, T, O, max_iter);
        extract_putative_key(B);
        float score = score_cipher(T);
        init_model(M, N, 0);
        g_random_seed = rand() % (1000000);
    }

    // init_ss_cipher_corpus(T);
    // Config: M = 26, N = 26, T = 1000

    // init_simple_corpus(T);
    // // Config: M = 3, N = 2, T = 4

    // init_brown_corpus(T);
    // // Config: M = 27, N = 2~, T = 50000, max_iter = 500

    // hmm_train(M, N, T, O, max_iter);

    
    // float score = score_cipher(T);
    // printf("score: %f\n", score);
    // printf("Fraction of putative key: %f and max score: %f\n", score, max_score);
    // if (score > max_score) max_score = score;

    // printf("MAX SCORE: %f\n", max_score);
    free_global_memory(N);
    return 0;
}

void hmm_train(int M, int N, int T, int* O, int max_iter) {
    printf("Start HMM Training...\n");
    // check input matrix
    for (int i = 0; i < N; i++) {
        double sum = 0.0;
        for (int j = 0; j < M; j++) sum += B[i][j];
        if (fabs(sum - 1.0) > 10e-7) { fprintf(stderr, "ERROR: B matrix is not stochastic (%f)\n", sum); exit(1); };

        sum = 0.0;
        for (int j = 0; j < N; j++) sum += A[i][j];
        if (fabs(sum - 1.0) > 10e-7) { fprintf(stderr, "ERROR: A matrix is not stochastic (%f)\n", sum); exit(1); };
    }

    dump_model(M, N);

    double old_prob = -(1.0 / 0.0); // to make it -INF
    double new_prob = 1.0;
    int iteration = 0;
    while (iteration < max_iter && new_prob > old_prob) {
        alpha_pass(N, T, O);
        beta_pass(N, T, O);
        compute_gamma(N, T, O);
        re_estimate(N, M, T, O);
        double new_prob = compute_prob(T);        
        old_prob = new_prob;
        iteration++;

        if (max_iter > 10 && (iteration % (int)(max_iter / 10.f)) == 0) {
            printf("----------------\n");
            printf("iter: %d(%.0f%%), prob: %f\n", iteration, iteration * 100.f / (float)max_iter, new_prob);
        }
    }
    dump_model(M, N);
}

void init_model(int M, int N, int update_A) {
    // A: NxN
    // B: NxM
    // pi: 1xN
    // Default is initializing to a nearly uniform distribution
    // buffer allocation
    A = (double **)malloc(sizeof(double*) * N);
    for (int i = 0; i < N; i++) A[i] = (double *)malloc(sizeof(double) * N);

    B = (double **)malloc(sizeof(double *) * N);
    for (int i = 0; i < N; i++) B[i] = (double *)malloc(sizeof(double) * M);

    pi = (double *)malloc(sizeof(double *) * N);

    // initialize value (nearly uniform distribution)
    srand(g_random_seed);
    const float precision = 1.f / 10.f;
    float cell[2] = {1.0 / N, 1.0 / M};
    float max_cell[2] = {0.f};
    float min_cell[2] = {0.f};
    float cell_range[2] = {0.f};
    for (int i = 0; i < 2; i++) {
        max_cell[i] = cell[i] + cell[i] * precision;
        min_cell[i] = cell[i] - cell[i] * precision;
        cell_range[i] = max_cell[i] - min_cell[i];
    }
    for (int i = 0; i < N; i++) {
        double sum = 0.0;
        if (update_A) {
            for (int j = 0; j < N - 1; j++) {
                float r = (rand() % 1000) / 1000.f;
                A[i][j] = min_cell[0] + cell_range[0] * r;
                sum += A[i][j];
            }
        }
        A[i][N - 1] = 1.0 - sum;

        sum = 0.0;
        for (int j = 0; j < M - 1; j++) {
            float r = (rand() % 1000) / 1000.f;
            B[i][j] = min_cell[1] + cell_range[1] * r;
            sum += B[i][j];
        }
        B[i][M - 1] = 1.0 - sum;

        float r = (rand() % 1000) / 1000.f;
        pi[i] = min_cell[0] + cell_range[0] * r;
    }
    pi[N - 1] = 1.0;
    for (int i = 0; i < N - 1; i++) pi[N - 1] -= pi[i];

    return;
}

void alpha_pass_no_scale(int N, int T, int *O) {
    // alpha is NxT
    if (!alpha) {
        alpha = (double **)malloc(sizeof(double *) * N);
        for (int i = 0; i < N; i++) alpha[i] = (double *)malloc(sizeof(double) * T);
    }

    if (!C) C = (double *)malloc(sizeof(double) * T);

    // compute alpha-pass
    for (int i = 0; i < N; i++) alpha[i][0] = pi[i] * B[i][O[0]];

    // compute alpha_t(i)
    for (int t = 1; t < T; t++) {
        for (int i = 0; i < N; i++) {
            alpha[i][t] = 0;
            for (int j = 0; j < N; j++) alpha[i][t] += (alpha[j][t - 1] * A[j][i]);
            alpha[i][t] *= B[i][O[t]];
        }
    }
}

void alpha_pass(int N, int T, int *O) {
    // alpha is NxT
    if (!alpha) {
        alpha = (double **)malloc(sizeof(double *) * N);
        for (int i = 0; i < N; i++) alpha[i] = (double *)malloc(sizeof(double) * T);
    }

    if (!C) C = (double *)malloc(sizeof(double) * T);

    // compute alpha-pass
    C[0] = 0.0;
    for (int i = 0; i < N; i++) {
        int q = O[0];
        alpha[i][0] = pi[i] * B[i][q];
        C[0] = C[0] + alpha[i][0];
    }

    // scale the alpha_0(i)
    C[0] = 1.0 / C[0];
    for (int i = 0; i < N; i++) alpha[i][0] *= C[0];

    // compute alpha_t(i)
    for (int t = 1; t < T; t++) {
        C[t] = 0;
        for (int i = 0; i < N; i++) {
            alpha[i][t] = 0;
            for (int j = 0; j < N; j++) {
                alpha[i][t] += (alpha[j][t - 1] * A[j][i]);
            }
            int q = O[t];
            alpha[i][t] *= B[i][q];
            C[t] += alpha[i][t];
        }

        // scale alpha_t(i)
        C[t] = 1.0 / C[t];
        for (int i = 0; i < N; i++) {
            alpha[i][t] *= C[t];
        }
    }
}

void beta_pass(int N, int T, int *O) {
    // beta is NxT
    if (!beta) {
        beta = (double **)malloc(sizeof(double *) * N);
        for (int i = 0; i < N; i++) beta[i] = (double *)malloc(sizeof(double) * T);
    }

    for (int i = 0; i < N; i++) {
        beta[i][T - 1] = C[T - 1];
    }

    for (int t = T - 2; t >= 0; t--) {
        for (int i = 0; i < N; i++) {
            beta[i][t] = 0.0;
            for (int j = 0; j < N; j++) {
                int q = O[t + 1];
                beta[i][t] += (A[i][j] * B[j][q] * beta[j][t + 1]);
            }
            // scale beta_t(i)
            beta[i][t] *= C[t];
        }
    }
}

void compute_gamma(int N, int T, int *O) {
    // gamma: NxT
    if (!gamma_t) {
        gamma_t = (double **)malloc(sizeof(double *) * N);
        for (int i = 0; i < N; i++) gamma_t[i] = (double *)malloc(sizeof(double) * T);
    }

    // // di-gamma: NxNxT
    if (!di_gamma) {
        di_gamma = (double ***)malloc(sizeof(double **) * N);
        for (int i = 0; i < N; i++) {
            di_gamma[i] = (double **)malloc(sizeof(double *) * N);
            for (int j = 0; j < N; j++) di_gamma[i][j] = (double *)malloc(sizeof(double) * T);
        }
    }

    for (int t = 0; t < T - 1; t++) {
        double denom  = 0.0;
        for (int i = 0; i < N; i++) {
#if 1 // they are should be equalvent. For performance concern, use faster one
            denom += alpha[i][T - 1];
#else
            for (int j = 0; j < N; j++) {
                int q = O[t + 1];
                denom += (alpha[i][t] * A[i][j] * B[j][q] * beta[j][t + 1]);
            }
#endif
        }

        for (int i = 0; i < N; i++) {
            gamma_t[i][t] = 0.0;
            for (int j = 0; j < N; j++) {
                int q = O[t + 1];
                di_gamma[i][j][t] = (alpha[i][t] * A[i][j] * B[j][q] * beta[j][t + 1]) / denom;
                gamma_t[i][t] += di_gamma[i][j][t];
            }
        }
    }

    // special case for gamma_t-1(i)
    double denom = 0.0;
    for (int i = 0; i < N; i++) {
        denom += alpha[i][T - 1];
    }

    for (int i = 0; i < N; i++) {
        gamma_t[i][T - 1] = alpha[i][T - 1] / denom;
    }
}

void re_estimate(int N, int M, int T, int *O) {
    // re-estimate pi
    for (int i = 0; i < N; i++) pi[i] = gamma_t[i][0];

    // // re-estimate A
    // for (int i = 0; i < N; i++) {
    //     for (int j = 0; j < N; j++) {
    //         double numer = 0.0;
    //         double denom = 0.0;
    //         for (int t = 0; t < T - 1; t++) {
    //             numer += di_gamma[i][j][t];
    //             denom += gamma_t[i][t];
    //         }
    //         A[i][j] = numer / denom;
    //     }
    // }

    // re-estimate B
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            double numer = 0.0;
            double denom = 0.0;
            for (int t = 0; t < T; t++) {
                if (O[t] == j) numer += gamma_t[i][t];
                denom += gamma_t[i][t];
            }
            B[i][j] = numer / denom;
        }
    }
}

double compute_prob(int T) {
    double log_prob = 0.0;
    for (int i = 0; i < T; i++) {
        log_prob += log(C[i]);
    }
    log_prob = -log_prob;
    return log_prob;
}

File: src/python/hmm.py
import re
import random
import math

if __name__ == "__main__":
    print("HMM testing itself.")


def train(file=None, *args):
    """
    Hidden Markov Model training entrance.
    :param file: file path of the corpus
    :param args: # of hidden states(N), # of observed symbols(M), # of observations(M), PRNG seed, max iteration
    :return:
    """
    if file is None or len(args) < 5:
        print("Parameters: file_path, N, M, T, seed, maxIter")
        return

    n_len = args[0]
    m_len = args[1]
    t_len = args[2]
    seed = args[3]
    max_iteration = args[4]

    corpus = [0] * t_len
    read_letters(file, t_len, corpus, True)

    a_matrix = [([0] * n_len) for i in range(n_len)]
    b_matrix = [([0] * m_len) for i in range(n_len)]
    pi_matrix = [0] * n_len
    model = {'A': a_matrix, 'B': b_matrix, 'pi': pi_matrix}
    observation = [0] * t_len

    init_model(n_len, m_len, model, seed)
    init_observation(observation, corpus)
    hmm_train(n_len, m_len, t_len, model, observation, max_iteration)


def init_observation(observation, corpus):
    """
    Initialize the observation in sense of english alphabet.
    Only lower case letter and space are treated as 27
    :param observation: observation list with length T
    :param corpus: list of all characters
    :return:
    """
    for i in range(len(observation)):
        if corpus[i] == ' ':
            observation[i] = 26
        else:
            observation[i] = ord(corpus[i]) - ord('a')


def read_letters(file, t_len, corpus, read_space=False):
    """
    Read alphabet from file to corpus
    :param file: file path
    :param t_len: length of corpus or length of observation
    :param corpus: output corpus list of chars
    :param read_space: a flag to read space or not
    :return:
    """
    count = 0
    with open(file) as f:
        for line in f:
            words = " ".join(re.findall("[a-zA-Z]+", line))
            for ch in words:
                if read_space is True and ch != ' ':
                    ch = ch.lower()
                corpus[count] = ch
                count = count + 1
                if count >= t_len:
                    break
            if count >= t_len:
                break


def init_model(n_len, m_len, model, seed):
    """
    Initialize lambda model with nearly an uniform distribution.
    Model = [A, B, pi]
    A: N x N
    B: N x M
    pi: 1 x N
    :param n_len: length of hidden states
    :param m_len: length of different observation symbols
    :param model: a dictionary with {'A': a_matrix, 'B': b_matrix, 'pi': pi_matrix}
    :param seed: initial random seed
    :return:
    """
    a = model['A']
    b = model['B']
    pi = model['pi']

    random.seed(seed)

    ceil_a = 1 / n_len
    ceil_b = 1 / m_len
    precision = 10

    for i in range(n_len):
        prob_sum = 0
        for j in range(n_len - 1):
            r = random.uniform(0, 1)
            if j % 2 == 0:
                a[i][j] = ceil_a + r * ceil_a / precision
            else:
                a[i][j] = ceil_a - r * ceil_a / precision
            prob_sum += a[i][j]
        a[i][n_len - 1] = 1 - prob_sum

        prob_sum = 0
        for j in range(m_len - 1):
            r = random.uniform(0, 1)
            if j % 2 == 0:
                b[i][j] = ceil_b + r * ceil_b / precision
            else:
                b[i][j] = ceil_b - r * ceil_b / precision
            prob_sum += b[i][j]
        b[i][m_len - 1] = 1 - prob_sum

    prob_sum = 0
    for i in range(n_len - 1):
        r = random.uniform(0, 1)
        if i % 2 == 0:
            pi[i] = ceil_a + r * ceil_a / precision
        else:
            pi[i] = ceil_a - r * ceil_a / precision
        prob_sum += pi[i]
    pi[n_len - 1] = 1 - prob_sum


def hmm_train(n_len, m_len, t_len, model, observation, max_iteration):
    """
    Kernel training entrance.
    Step1: alpha_pass
    Step2: beta_pass
    Step3: gamma_pass
    Step4: re_estimate
    Step5: evaluate model
    Step6: Goto step1 if needed
    :param n_len: length of hidden states
    :param m_len: length of different observed symbols
    :param t_len: length of observation sequence
    :param model: lambda model
    :param observation: list of observation sequence
    :param max_iteration: max iteration of HMM
    :return:
    """
    for i in range(n_len):
        test_sum = 0
        for j in range(m_len):
            test_sum += model['B'][i][j]
        print("sum = ", test_sum)

    scaling_factor = [0] * t_len
    alpha = [([0] * t_len) for i in range(n_len)]
    beta = [([0] * t_len) for i in range(n_len)]
    gamma = [([0] * t_len) for i in range(n_len)]
    di_gamma = [([([0] * t_len) for i in range(n_len)]) for i in range(n_len)]

    dump_model(n_len, m_len, model)

    old_prob = 0
    for i in range(max_iteration):
        alpha_pass(n_len, t_len, observation, alpha, model, scaling_factor)
        beta_pass(n_len, t_len, observation, beta, model, scaling_factor)
        gamma_pass(n_len, t_len, observation, alpha, beta, gamma, di_gamma, model)
        re_estimate(n_len, m_len, t_len, observation, gamma, di_gamma, model)
        log_prob = get_probability(t_len, scaling_factor)
        if i > 1 and log_prob < old_prob:
            break
        old_prob = log_prob
        print("log_prob: ", log_prob)
    dump_model(n_len, m_len, model)


def alpha_pass(n_len, t_len, observation, alpha, model, c):
    """
    HMM training step1: Forward Algorithm(alpha pass).
    :param n_len: length of hidden states
    :param t_len: length of observation sequence
    :param observation: list of observation
    :param alpha: output alpha list with size N x T
    :param model: HMM model
    :param c: scaling factor with size 1 x T
    :return:
    """
    a = model['A']
    b = model['B']
    pi = model['pi']
    c[0] = 0
    for i in range(0, n_len):
        alpha[i][0] = pi[i] * b[i][observation[0]]
        c[0] += alpha[i][0]

    c[0] = 1 / c[0]
    for i in range(0, n_len):
        alpha[i][0] *= c[0]

    for t in range(1, t_len):
        c[t] = 0
        for i in range(0, n_len):
            alpha[i][t] = 0
            for j in range(0, n_len):
                alpha[i][t] += alpha[j][t - 1] * a[j][i]
            alpha[i][t] *= b[i][observation[t]]
            c[t] += alpha[i][t]

        c[t] = 1.0 / c[t]
        for i in range(0, n_len):
            alpha[i][t] *= c[t]


def beta_pass(n_len, t_len, observation, beta, model, c):
    """
    HMM training step2: Backward Algorithm(beta pass)
    :param n_len: length of hidden states
    :param t_len: length of observation sequence
    :param observation: list of observation sequence
    :param beta: output beta list of size N x M
    :param model: HMM model
    :param c: scaling factor with size 1 x T
    :return:
    """
    a = model['A']
    b = model['B']
    for i in range(0, n_len):
        beta[i][t_len - 1] = c[t_len - 1]

    for t in range(t_len - 2, -1, -1):
        for i in range(0, n_len):
            beta[i][t] = 0
            for j in range(0, n_len):
                beta[i][t] += a[i][j] * b[j][observation[t + 1]] * beta[j][t + 1]
            beta[i][t] *= c[t]


def gamma_pass(n_len, t_len, observation, alpha, beta, gamma, di_gamma, model):
    """
    HMM training step3: Compute gamma and di_gamma.
    :param n_len: length of hidden states
    :param t_len: length of observation sequence
    :param observation: list of observation sequence
    :param alpha: the alpha vector from alpha bass
    :param beta: the beta vector from beta pass
    :param gamma: output gamma list with size N x T
    :param di_gamma: output di_gamma list with size N x N x T
    :param model: HMM model
    :return:
    """
    a = model['A']
    b = model['B']
    for t in range(t_len - 1):
        denom = 0
        for i in range(n_len):
            denom += alpha[i][t_len - 1];

        for i in range(n_len):
            gamma[i][t] = 0
            for j in range(n_len):
                di_gamma[i][j][t] = alpha[i][t] * a[i][j] * b[j][observation[t + 1]] * beta[j][t + 1] / denom
                gamma[i][t] += di_gamma[i][j][t]

    denom = 0
    for i in range(n_len):
        denom += alpha[i][t_len - 1]
    for i in range(n_len):
        gamma[i][t_len - 1] = alpha[i][t_len - 1] / denom


def re_estimate(n_len, m_len, t_len, observation, gamma, di_gamma, model):
    """
    HMM training step4: Re-estimate model A, B, and pi matrix.
    :param n_len: length of hidden states
    :param m_len: length of different observed symbols
    :param t_len: length of observation sequence
    :param observation: list of observation sequence
    :param gamma: the gamma vector from gamma_pass
    :param di_gamma: the di_gamma vector from gamma_pass
    :param model: HMM model
    :return:
    """
    a = model['A']
    b = model['B']
    pi = model['pi']

    for i in range(n_len):
        pi[i] = gamma[i][0]

    for i in range(n_len):
        for j in range(n_len):
            numer = 0
            denom = 0
            for t in range(t_len - 1):
                numer += di_gamma[i][j][t]
                denom += gamma[i][t]
            a[i][j] = numer / denom

        for k in range(m_len):
            numer = 0
            denom = 0
            for t in range(t_len):
                if observation[t] == k:
                    numer += gamma[i][t]
                denom += gamma[i][t]
            b[i][k] = numer / denom


def get_probability(t_len, scaling):
    """
    HMM training step5: Evaluate new observation probability given an HMM model
    :param t_len: length of observation sequence
    :param scaling: scaling factor
    :return: logarithm probability
    """
    log_prob = 0
    for i in range(t_len):
        log_prob += math.log(scaling[i])
    log_prob = -log_prob
    return log_prob


def dump_model(n_len, m_len, model):
    """
    Print the model values for debug.
    :param n_len: length of hidden states
    :param m_len: length of different observations
    :param model: HMM model
    :return:
    """
    print("\n----A matrix----")
    for i in range(n_len):
        for j in range(n_len): print("%.4f" % model['A'][i][j], " ", end="")
        print()
    print("\n----B matrix----")
    for i in range(n_len):
        for j in range(m_len): print("%.4f" % model['B'][i][j], " ", end="")
        print()
    print("\n----pi matrix----")
    for i in range(n_len): print("%.4f" % model['pi'][i], " ", end="")
    print()